MQE-987: Decouple MFTF from Magento

imeron2433 opened the pull request:
- add command parity to mftf bin command
- provide support for local bin execution in funcitonal tests

<!--- Provide a general summary of the Pull Request in the Title above -->

### Description
<!--- Provide a description of the changes proposed in the pull request -->

### Fixed Issues (if relevant)
<!--- Provide a list of fixed issues in the format magento/magento2-functional-testing-framework#<issue_number>, if relevant  -->
1. magento/magento2-functional-testing-framework#<issue_number>: Issue title
2. ...

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/verification tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)
 - [ ] Changes to Framework doesn't have backward incompatible changes for tests or have related Pull Request with fixes to tests

okolesnyk commented:
@lenaorobei please review

imeron2433 commented on dev/tests/functional/_bootstrap.php:
```diff
@@ -29,7 +16,17 @@
+    foreach ($_ENV as $key => $var) {
+        defined($key) || define($key, $var);
+    }
+}
```
fixed.

imeron2433 commented on src/Magento/FunctionalTestingFramework/Console/CleanProjectCommand.php:
```diff
@@ -0,0 +1,109 @@
+
+        return $exists;
+    }
+}
```
Done.

imeron2433 commented on src/Magento/FunctionalTestingFramework/Console/RunTestCommand.php:
```diff
@@ -0,0 +1,73 @@
+        $this->setName("run:test")
+            ->setDescription("generation and execution of test(s) defined in xml")
+            ->addArgument('name', InputArgument::REQUIRED | InputArgument::IS_ARRAY, "name of tests to generate and execute")
+            ->addOption('skip-generate', 'k', InputOption::VALUE_NONE, "skip generation and execute existing test");
```
I used k because usually the argument is associated with a "silent" execution and the shortcut must be one letter.

imeron2433 commented on src/Magento/FunctionalTestingFramework/Console/RunTestCommand.php:
```diff
@@ -0,0 +1,73 @@
+        );
+    }
+
+}
```
Done

imeron2433 commented on src/Magento/FunctionalTestingFramework/_bootstrap.php:
```diff
@@ -0,0 +1,62 @@
+);
+
+// add the debug flag here
+$debug_mode = $_ENV['MFTF_DEBUG'] ?? false;
```
if we do not coalesce and variable is not set we will get an "undefined array key" exception. I will change the variable name to be camelCase.

imeron2433 commented on src/Magento/FunctionalTestingFramework/Console/CleanProjectCommand.php:
```diff
@@ -0,0 +1,109 @@
+        if ($output->isVerbose()) {
+            // build a list of the paths
+            $afterState = $this->getFileSystemState($finder);
+            $deletedFiles = array_diff($beforeState, $afterState);
```
I am not sure I understand this comment. From what I understand if a file does not exist and filesystem attempts to remove there is no exception, how can I rely on this list? Additionally I am deleting php files in certain directories, that will not tell me explicit file deletion to output.

lenaorobei replied:
Maybe I'm missing something, but this is how I understood it should work.
```
protected function execute(InputInterface $input, OutputInterface $output)
    {
        $fileSystem = new Filesystem();

        // delete the files mftf generates during test execution in TESTS_BP
        if (is_dir(self::GENERATED_FILES)) {
            $fileSystem->remove(self::GENERATED_FILES);
            if ($output->isVerbose()) {
                $output->writeln("Folder with generated files deleted.");
            }
        }
        
        $finder = new Finder();
        $files = $finder->files()->name('*.php')->in(FW_BP . '/src/Magento/FunctionalTestingFramework/Group/');
        foreach ($files as $file) {
            $path = $file->getRealPath();
            $fileSystem->remove($file);
            if ($output->isVerbose()) {
                $output->writeln(sprintf('File %s deleted.', $path));
            }
        }

        // delete config files if user specifies a hard reset
        $isHardReset = $input->getOption('hard');
        if ($isHardReset) {
            foreach (self::CONFIGURATION_FILES as $configFile) {
                if (file_exists($configFile)) {
                    $fileSystem->remove($configFile);
                    if ($output->isVerbose()) {
                        $output->writeln("Config file $configFile deleted.");
                    }
                }
            }
        }

        $output->writeln('<info>Project cleaned successfully.<info>');
    }
```

imeron2433 replied:
Okay I see what you are saying, I didn't want to explicitly loop the files since the filesystem command will go ahead and delete the whole group on its own but you're right if we want to have the explicit print we'll have to loop the files ourselves and output.

imeron2433 commented on src/Magento/FunctionalTestingFramework/Console/GenerateTestsCommand.php:
```diff
@@ -0,0 +1,140 @@
+        $this->setName('generate:tests')
+            ->setDescription('This command generates all test files and suites based on xml declarations')
+            ->addArgument('name', InputArgument::OPTIONAL | InputArgument::IS_ARRAY, 'name(s) of specific tests to generate')
+            ->addOption("config", null, InputOption::VALUE_REQUIRED, 'default, singleRun, or parallel', 'default')
```
Done

imeron2433 commented on src/Magento/FunctionalTestingFramework/Console/RunTestCommand.php:
```diff
@@ -0,0 +1,73 @@
+        $process->setWorkingDirectory(TESTS_BP);
+        $process->run(
+            function ($type, $buffer) use ($output) {
+                $output->write($buffer);
```
Done

imeron2433 commented on src/Magento/FunctionalTestingFramework/Console/RunTestGroupCommand.php:
```diff
@@ -0,0 +1,104 @@
+     */
+    private function getGroupAndSuiteConfiguration(array $groups)
+    {
+        $testConfiguration['tests'] = [];
```
I would like them both to be null, but I cannot merge any relevant tests to a value of null. I could do a check for the "first value" and replace instead of merge but this seems like elegant to me. Additionally I could make this command exclusive to test groups and point users to the run suite command if they attempt to run a group that is in fact a suite.

imeron2433 commented on src/Magento/FunctionalTestingFramework/Console/GenerateSuiteCommand.php:
```diff
@@ -0,0 +1,56 @@
+        }
+
+        if ($output->isVerbose()) {
+            $output->writeLn("All Suites Generated");
```
Done.

imeron2433 commented:
Hi @lenaorobei, I believe i have addressed all of your comments. I am going to merge this code in at the End of the Day if you don't have any other concerns.
